File: file_via_socket.py
# This is the server side script for receiving data via IP socket. Creates a file for each socket connection,
# with timestamp in the file name.
# For details see the GitHub repository https://github.com/viktor-nikolov/lwIP-file-via-socket
#
# Run the script with the command 'python3 file_via_socket [params]' or 'python file_via_socket [params]'.
# Tested on Ubuntu 22.04 and Windows 11.
#
# usage: file_via_socket [-h] [--path PATH] [--prefix PREFIX] [--ext EXT] [--bind_ip BIND_IP] [--bind_port BIND_PORT]
#
# options:
#   -h, --help              Show help message and exit
#   --path PATH             Path for storing the files; defaults to current working directory
#   --prefix PREFIX         Prefix of the file name; defaults to "via_socket"
#   --ext EXT               Extension for the file name; defaults to "txt"
#   --bind_ip BIND_IP       Local IP to bind the socket listener to; defaults to 0.0.0.0
#   --bind_port BIND_PORT   Local port to listen to connection on; values: 1024..65535; defaults to 65432
#
# BSD 2-Clause License:
#
# Copyright (c) 2024 Viktor Nikolov
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
# 1. Redistributions of source code must retain the above copyright notice, this
#    list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright notice,
#    this list of conditions and the following disclaimer in the documentation
#    and/or other materials provided with the distribution.
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

import ipaddress
import socket
import signal
import argparse
from datetime import datetime
from select import select


def signal_handler(signum, frame):
    # Handler for Ctlr+C signal
    print("\nExecution interrupted by the user. Exiting...")
    exit(1)


def check_valid_ip(arg_val):
    # Checking value of IP address passed as command line argument
    # We attempt to create an IPv4 or IPv6 address object. If successful, the string is a valid IP address.
    # Otherwise, ValueError exception is raised.
    ipaddress.ip_address(arg_val)
    return arg_val


def check_port_value(arg_val):
    # Checking value of port passed as command line argument
    port = int(arg_val)
    if port <= 1023 or port > 65535:
        raise ValueError()
    return port


def bytes2human_readable(bytes_val):
    # Convert number of bytes to a human-readable value with units
    if bytes_val > 1024 * 1024 * 1024:
        return "{:.2f} GB".format(bytes_val / (1024 * 1024 * 1024))
    elif bytes_val > 1024 * 1024:
        return "{:.2f} MB".format(bytes_val / (1024 * 1024))
    elif bytes_val > 1024:
        return "{:.2f} KB".format(bytes_val / 1024)
    else:
        return "{:} B".format(bytes_val)


signal.signal(signal.SIGINT, signal_handler)  # Handling Ctrl+C

# Global variables
filePath = ""
filePrefix = "via_socket"
fileExt = "txt"
bindIP = "0.0.0.0"  # Standard loopback interface address (localhost)
bindPort = 65432    # Port to listen on (non-privileged ports are > 1023)
fileName = ""       # Initializing the variable to avoid a warning

# Create the command line argument parser
parser = argparse.ArgumentParser(prog="file_via_socket",
                                 description='Server side script for receiving data via IP socket. '
                                             'Creates a file for each socket connection, with timestamp in the file name.')
# Add arguments
parser.add_argument('--path', type=str, help='path for storing the files; defaults to current working directory')
parser.add_argument('--prefix', type=str, help=f'prefix of the file name; defaults to "{filePrefix}"')
parser.add_argument('--ext', type=str, help=f'extension for the file name; defaults to "{fileExt}"')
parser.add_argument('--bind_ip', type=check_valid_ip, help=f'local IP to bind the socket listener to; defaults to {bindIP}')
parser.add_argument('--bind_port', type=check_port_value,
                    help=f'local port to listen to connection on; values: 1024..65535; defaults to {bindPort}')
# Parse the command line arguments
args = parser.parse_args()
# Store command line arguments values
if args.path is not None:
    filePath = args.path.rstrip('/\\')  # remove trailing character '/' or '\'
if args.prefix is not None:
    filePrefix = args.prefix
if args.ext is not None:
    fileExt = args.ext.lstrip('.')  # remove leading character '.'
if args.bind_ip is not None:
    bindIP = args.bind_ip
if args.bind_port is not None:
    bindPort = args.bind_port

print(f"Waiting for connection on {bindIP}:{bindPort}\n"
       "(Press Ctrl+C to terminate)")

with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
    #  Set the IP socket
    s.bind((bindIP, bindPort))
    s.listen(1)
    while True:  # The outer cycle is for each socket connection, i.e., for each file
        while True:
            # We are using "select" because if we called s.accept() directly, it wouldn't be terminated by Ctrl+C
            ready, _, _ = select([s], [], [], 1)  # Check if we have an incoming connection; timeout set to 1 sec
            if ready:
                conn, addr = s.accept()  # Connection object is returned
                break  # We have a connection, let's handle incoming data

        # Construct the file name
        fileName = ""
        if filePath != "":
            fileName = filePath + '/'
        now = datetime.now()
#                                             date and time             first 4 digits of microseconds
        fileName += filePrefix + now.strftime("_%y%m%d_%H%M%S") + '.' + now.strftime('%f')[:4]
        if fileExt != "":
            fileName += "." + fileExt
        try:
            with conn, open(fileName, 'wb') as f:
                print(f"Got connection from {addr[0]}:{addr[1]}")
                dataTotal = 0
                while True:  # Reading data sent by client via the socket
                    try:
                        data = conn.recv(4096)
                        if data:
                            f.write(data)
                            dataTotal += len(data)
                        if not data:  # Empty data means that the client has disconnected
                            break
                    except ConnectionResetError:
                        break
                print(f"    Received total: {bytes2human_readable(dataTotal)}")
        except FileNotFoundError:
            print(f"ERROR: Unable to open file '{fileName}'")
            exit(1)
        except PermissionError:
            print(f"ERROR: Permission error when opening file '{fileName}'")
            exit(1)

File: demo_app_FreeRTOS_on_Zynq/project_files/DemoFileViaSocket_sw/DemoFileViaSocket/src/FileViaSocket.cpp
/*
This is the source file for the C++ ostream class for writing a file on a remote system via an IP socket connection.
Details are explained on GitHub: https://github.com/viktor-nikolov/lwIP-file-via-socket

Tested (and ready for compilation) on FreeRTOS on AMD Xilinx Zynq SoC (Vitis 2023.1 toolchain),
Windows 11 (MinGW toolchain) and Ubuntu 22.04 (gcc toolchain).

BSD 2-Clause License:

Copyright (c) 2024 Viktor Nikolov

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:
1. Redistributions of source code must retain the above copyright notice, this
   list of conditions and the following disclaimer.
2. Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation
   and/or other materials provided with the distribution.
THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
#include "FileViaSocket.h"
#include <cstring>

#ifdef __WIN32__
#   include <winsock2.h>
#   define SHUTDOWN_HOW_BOTH SD_BOTH   // We pass this as a parameter to function shutdown()
#elif defined(__linux__)
#   include <sys/socket.h>
#   include <arpa/inet.h>
#   include <unistd.h>
#   define SHUTDOWN_HOW_BOTH SHUT_RDWR // We pass this as a parameter to function shutdown()
#else // If not Windows nor Linux, we assume FreeRTOS with lwIP
/* When using lwIP and ostream together, we face an issue with two different definitions of errno.
 * lwIP defines errno as a global variable in lwip/errno.h. lwIP functions set this global variable.
 * However, when we include header ostream it takes with it sys/errno.h
 * (for example Xilinx/Vitis/2023.1/gnu/aarch32/nt/gcc-arm-none-eabi/aarch32-xilinx-eabi/usr/include/errno.h),
 * which defines errno differently (as an attribute of global reentrancy structure, returned by function).
 * By un-defining macro "errno" defined in sys/errno.h, we get access to the global variable errno defined
 * in  lwip/errno.h. */
#undef errno

/* Following macros are defined in lwip/errno.h differently from sys/errno.h. We are un-defining
 * them to avoid compiler warning (there is no other way to disable this warning).
 * sys/errno.h was included from within header ostream. */
#   undef ENAMETOOLONG
#   undef EDEADLK
#   undef ENOLCK
#   undef ENOSYS
#   undef ENOTEMPTY
#   undef ELOOP
#   undef ENOMSG
#   undef EIDRM
#   undef EMULTIHOP
#   undef EBADMSG
#   undef EOVERFLOW
#   undef EILSEQ
#   undef ENOTSOCK
#   undef EDESTADDRREQ
#   undef EMSGSIZE
#   undef EPROTOTYPE
#   undef EPROTONOSUPPORT
#   undef EAFNOSUPPORT
#   undef EADDRINUSE
#   undef EADDRNOTAVAIL
#   undef ENETDOWN
#   undef ENETUNREACH
#   undef ENETRESET
#   undef ECONNABORTED
#   undef EISCONN
#   undef ENOTCONN
#   undef ETOOMANYREFS
#   undef ETIMEDOUT
#   undef EHOSTDOWN
#   undef EHOSTUNREACH
#   undef EALREADY
#   undef EINPROGRESS
#   undef ESTALE
#   undef EDQUOT
#   undef ENOPROTOOPT

#   include "lwip/sockets.h"
#   define SHUTDOWN_HOW_BOTH SHUT_RDWR  // We pass this as a parameter to function shutdown()
#   define INADDR_NONE IPADDR_NONE      // lwIP doesn't provide macro INADDR_NONE
#   undef close  // Macro "close" defined in lwip/sockets.h messes with our methods named "close"
#endif

void SocketBuffer::open( const std::string &serverIP, unsigned short port )
{
	if( Socket >= 0 ) { // We still have an open socket from before
		close();
		Socket = -1;
	}

	// Create socket
	if( (Socket = socket(AF_INET, SOCK_STREAM, 0 )) < 0 )
#ifdef __WIN32__
		throw FileViaSocket::SocketCreationErrorExc( WSAGetLastError() );
#else
		throw FileViaSocket::SocketCreationErrorExc( errno );
#endif

	struct sockaddr_in serv_addr = {};
	serv_addr.sin_family = AF_INET;
	serv_addr.sin_port = htons(port);

	// Convert IPv4 and IPv6 addresses from text to binary form
	serv_addr.sin_addr.s_addr = inet_addr( serverIP.c_str() );
	if( serv_addr.sin_addr.s_addr == INADDR_NONE ) {
		std::string m{"Server IP was provided in a wrong format '" + serverIP + "'!"};
		throw FileViaSocket::WrongServerIPFormatExc( m );
	}

	// Connect to the server
	if( connect(Socket, (struct sockaddr *)&serv_addr, sizeof(serv_addr)) < 0 )
#ifdef __WIN32__
		throw FileViaSocket::SocketConnectionErrorExc( WSAGetLastError() );
#else
		throw FileViaSocket::SocketConnectionErrorExc( errno );
#endif
} // SocketBuffer::open

void SocketBuffer::close()
{
	if( Socket >= 0 ) {
		sync(); // Write remaining data from the buffer to the socket
		shutdown(Socket, SHUTDOWN_HOW_BOTH); // Gracefully closing the socket

#ifdef __WIN32__
		closesocket( Socket ); // Calling Winsock2 function for closing the socket
#elif defined(__linux__)
		::close( Socket );     // Calling the global library function for closing the file descriptor
#else // If not Windows nor Linux, we assume FreeRTOS with lwIP
		lwip_close( Socket );
#endif
		Socket = -1;
	}
} // SocketBuffer::close

int SocketBuffer::overflow( int c ) {
	if( Socket < 0 )
		return traits_type::eof();

	if ( c == traits_type::eof() ) {
		sync();
		return 1; // Success
	}

	if( bytesInBuffer == SOCKET_BUFF_SIZE-1 ) { // This character fills the buffer
		buffer[ bytesInBuffer ] = char(c);

		if( send(Socket, buffer, SOCKET_BUFF_SIZE, 0) != SOCKET_BUFF_SIZE )
			return traits_type::eof(); // Failure
		bytesInBuffer = 0;
	}
	else { // There is space in the buffer for more characters
		buffer[ bytesInBuffer ] = char(c);
		bytesInBuffer++;
	}

	return 1; // Success
} // SocketBuffer::overflow

std::streamsize SocketBuffer::xsputn( const char_type* s, std::streamsize n )
{
	if( Socket < 0 )
		return 0; // Failure

	if( bytesInBuffer + n >= SOCKET_BUFF_SIZE ) { // Data won't fit in the buffer; we need to send data to the socket
		std::streamsize bytesConsumed{0}; // Number of bytes we already consumed form s

		// If we have data in the buffer, we fill the buffer to be full and send it
		if( bytesInBuffer > 0 ) {
			memcpy( buffer + bytesInBuffer, s, SOCKET_BUFF_SIZE - bytesInBuffer );
			bytesConsumed = SOCKET_BUFF_SIZE - bytesInBuffer;
			if( send(Socket, buffer, SOCKET_BUFF_SIZE, 0) != SOCKET_BUFF_SIZE )
				return 0; // Failure
		}

		// Now send all data from s, which would not fit in the buffer
		std::streamsize n2 = ( n - bytesConsumed ) - ( n - bytesConsumed ) % SOCKET_BUFF_SIZE;
		if( n2 > 0 ) { // Is there something to send?
			if( send(Socket, s + bytesConsumed, n2, 0) != n2 )
				return bytesConsumed; // Failure
			bytesConsumed += n2;
		}

		if( bytesConsumed < n ) { // We store data, which are still remaining, in the buffer
			memcpy( buffer, s + bytesConsumed, n - bytesConsumed );
			bytesInBuffer = n - bytesConsumed;
		} else {
			bytesInBuffer = 0; // Otherwise, we are left with an empty buffer
		}
	}
	else { // Data still fit in the buffer
		memcpy( buffer + bytesInBuffer, s, n );
		bytesInBuffer += n;
	}
	return n; // Success
} // SocketBuffer::xsputn

int SocketBuffer::sync()
{
	if( bytesInBuffer == 0 ) // No data to send
		return 0; // Success

	if( Socket < 0 )
		return -1; // Failure

	if( send(Socket, buffer, bytesInBuffer, 0) != bytesInBuffer )
		return -1; // Failure
	bytesInBuffer = 0;

	return 0; // Success
} // SocketBuffer::sync

FileViaSocket::SocketCreationErrorExc::SocketCreationErrorExc( int errCode )
{
#ifdef __WIN32__
	message = "Socket creation error! WSAGetLastError() == " + std::to_string(errCode);
#else
	message = "Socket creation error! errno == " + std::to_string(errno);
#endif
} // FileViaSocket::SocketCreationErrorExc

FileViaSocket::SocketConnectionErrorExc::SocketConnectionErrorExc( int errCode )
{
#ifdef __WIN32__
	message = "Socket connection error! WSAGetLastError() == " + std::to_string(errCode);
	if( errCode == WSAECONNREFUSED )
		message += " (connection refused; is server running?)";
	else if( errCode == WSAETIMEDOUT )
		message += " (connection timed out; is server accessible?)";
#else
	message = "Socket connection error! errno == " + std::to_string(errno);

	switch (errCode) {
		case ECONNREFUSED: // On Linux this error is raised when server is not running
			               // on the target host.
			message += " (connection refused; is server running?)";
			break;
		case ETIMEDOUT:
			message += " (connection timed out; is server accessible?)";
			break;
		case ECONNRESET:   // On lwIP this error is raised when server is not running
			               // on the target host.
			message += " (connection reset by peer; is server running?)";
			break;
		case ECONNABORTED: // On lwIP this error is raised instead of ETIMEDOUT
			message += " (SW caused connection abort; is server accessible?)";
			break;
	}
#endif
} // FileViaSocket::SocketConnectionErrorExc
